File: app/src/main/java/com/tpv/android/utils/audio/RecordingThread.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package com.tpv.android.utils.audio;

import android.media.AudioFormat;
import android.media.AudioRecord;
import android.media.MediaRecorder;
import android.util.Log;

import com.naman14.androidlame.AndroidLame;
import com.naman14.androidlame.LameBuilder;

import java.io.BufferedOutputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.LinkedList;

public class RecordingThread {
    private static final String TAG = "RecordingThread";
    private static final String LOG_TAG = RecordingThread.class.getSimpleName();
    private static final int SAMPLE_RATE = 8000;

    public RecordingThread(String filePath, AudioDataReceivedListener listener) {
        this.filePath = filePath;
        mListener = listener;
    }

    private boolean mShouldContinue;
    private String filePath;
    private AudioDataReceivedListener mListener;
    private Thread mThread;
    private BufferedOutputStream os = null;

    public boolean recording() {
        return mThread != null;
    }

    public void startRecording() {
        if (mThread != null)
            return;

        try {
            os = new BufferedOutputStream(new FileOutputStream(filePath));

        } catch (FileNotFoundException e) {
            Log.e(TAG, "File not found for recording ", e);
        }

        mShouldContinue = true;
        mThread = new Thread(new Runnable() {
            @Override
            public void run() {
                record();
            }
        });
        mThread.start();
    }

    public void stopRecording() {
        if (mThread == null)
            return;

        mShouldContinue = false;
        mThread = null;
    }

    private void record() {
        Log.v(LOG_TAG, "Start");
        android.os.Process.setThreadPriority(android.os.Process.THREAD_PRIORITY_AUDIO);

        // buffer size in bytes
        int bufferSize = AudioRecord.getMinBufferSize(SAMPLE_RATE,
                AudioFormat.CHANNEL_IN_MONO,
                AudioFormat.ENCODING_PCM_16BIT);

        if (bufferSize == AudioRecord.ERROR || bufferSize == AudioRecord.ERROR_BAD_VALUE) {
            bufferSize = SAMPLE_RATE * 2;
        }

        short[] audioBuffer = new short[bufferSize];
        LinkedList<short[]> audioBufferList = new LinkedList<>();
        LinkedList<Integer> bufferSizeList = new LinkedList<>();

        AudioRecord record = new AudioRecord(MediaRecorder.AudioSource.MIC,
                SAMPLE_RATE,
                AudioFormat.CHANNEL_IN_MONO,
                AudioFormat.ENCODING_PCM_16BIT,
                bufferSize);

        if (record.getState() != AudioRecord.STATE_INITIALIZED) {
            Log.e(LOG_TAG, "Audio Record can't initialize!");
            return;
        }
        record.startRecording();

        Log.v(LOG_TAG, "Start recording");


        AndroidLame androidLame = new LameBuilder()
                .setInSampleRate(SAMPLE_RATE)
                .setOutChannels(1)
                .setOutBitrate(32)
                .setOutSampleRate(SAMPLE_RATE)
                .build();


        long shortsRead = 0;
        while (mShouldContinue) {
            int numberOfShort = record.read(audioBuffer, 0, bufferSize);
            shortsRead += numberOfShort;

            audioBufferList.add(audioBuffer);
            bufferSizeList.add(numberOfShort);

            // Notify waveform
            mListener.onAudioDataReceived(audioBuffer);
        }


        byte[] byteAudio = new byte[(int) (7200 + bufferSize * 2.5)];


        for (int i = 0; i < audioBufferList.size(); i++) {
            int numberOfShort = bufferSizeList.get(i);
            short[] b = audioBufferList.get(i);
            try {
                int bytesEncoded = androidLame.encode(b, b, numberOfShort, byteAudio);
                os.write(byteAudio, 0, bytesEncoded);
            } catch (IOException e) {
                Log.e(TAG, "Error saving recording ", e);
                return;
            }
        }


        try {
            int outputMp3buf = androidLame.flush(byteAudio);
            os.write(byteAudio, 0, outputMp3buf);
            os.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
        record.stop();
        record.release();


        Log.v(LOG_TAG, String.format("Recording stopped. Samples read: %d", shortsRead));
    }
}

File: app/src/main/java/com/tpv/android/utils/audio/RecordingThread.kt
/*
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package com.tpv.android.utils.audio

import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import android.os.Message
import com.livinglifetechway.k4kotlin.core.orZero

import java.io.File
import java.io.IOException

class RecordingThread
/**
 * Default constructor. Setup recorder with default sampling rate 1 channel,
 * 16 bits pcm
 */(private val mRecordFile: File, private val mListener: AudioDataReceivedListener) {
    private var mAudioRecord: AudioRecord? = null
    private var mBufferSize: Int = 0
    private var mPCMBuffer: ShortArray? = null
    private var mEncodeThread: DataEncodeThread? = null
    var isRecording = false
        private set

    var volume: Int = 0
        private set

    val maxVolume: Int
        get() = MAX_VOLUME

    /**
     * Start recording. Create an encoding thread. Start record from this
     * thread.
     *
     * @throws IOException
     */
    @Throws(IOException::class)
    fun start() {
        if (isRecording)
            return
        initAudioRecorder()
        mAudioRecord?.startRecording()
        object : Thread() {

            override fun run() {
                // 设置线程权限
                android.os.Process
                        .setThreadPriority(android.os.Process.THREAD_PRIORITY_URGENT_AUDIO)
                isRecording = true
                while (isRecording) {
                    val readSize = mAudioRecord?.read(mPCMBuffer!!, 0, mBufferSize).orZero()
                    if (readSize > 0) {
                        mPCMBuffer?.let { mEncodeThread?.addTask(it, readSize) }
                        calculateRealVolume(mPCMBuffer, readSize)

                        mListener.onAudioDataReceived(mPCMBuffer)
                    }
                }
                // release and finalize audioRecord
                mAudioRecord?.stop()
                mAudioRecord?.release()
                mAudioRecord = null
                // stop the encoding thread and try to wait
                // until the thread finishes its job
                val msg = Message.obtain(mEncodeThread?.handler,
                        DataEncodeThread.PROCESS_STOP)
                msg.sendToTarget()
            }

            /**
             * 此计算方法来自samsung开发范例
             *
             * @param buffer
             * buffer
             * @param readSize
             * readSize
             */
            private fun calculateRealVolume(buffer: ShortArray?, readSize: Int) {
                var sum = 0
                for (i in 0 until readSize) {
                    // 这里没有做运算的优化，为了更加清晰的展示代码
                    sum += buffer?.get(i).orZero() * buffer?.get(i).orZero()
                }
                if (readSize > 0) {
                    val amplitude = (sum / readSize).toDouble()
                    volume = Math.sqrt(amplitude).toInt()
                }
            }
        }.start()
    }

    fun stop() {
        isRecording = false
    }

    /**
     * Initialize audio recorder
     */
    @Throws(IOException::class)
    private fun initAudioRecorder() {
        mBufferSize = AudioRecord.getMinBufferSize(DEFAULT_SAMPLING_RATE,
                DEFAULT_CHANNEL_CONFIG, DEFAULT_AUDIO_FORMAT.audioFormat)

        val bytesPerFrame = DEFAULT_AUDIO_FORMAT.bytesPerFrame
        /*
		 * Get number of samples. Calculate the buffer size (round up to the
		 * factor of given frame size) 使能被整除，方便下面的周期性通知
		 */
        var frameSize = mBufferSize / bytesPerFrame
        if (frameSize % FRAME_COUNT != 0) {
            frameSize += FRAME_COUNT - frameSize % FRAME_COUNT
            mBufferSize = frameSize * bytesPerFrame
        }

        /* Setup audio recorder */
        mAudioRecord = AudioRecord(DEFAULT_AUDIO_SOURCE,
                DEFAULT_SAMPLING_RATE, DEFAULT_CHANNEL_CONFIG,
                DEFAULT_AUDIO_FORMAT.audioFormat, mBufferSize)

        mPCMBuffer = ShortArray(mBufferSize)
        /*
		 * Initialize lame buffer mp3 sampling rate is the same as the recorded
		 * pcm sampling rate The bit rate is 32kbps
		 */
//        MP3Encoder.init(DEFAULT_SAMPLING_RATE, DEFAULT_LAME_IN_CHANNEL,
//                DEFAULT_SAMPLING_RATE, DEFAULT_LAME_MP3_BIT_RATE,
//                DEFAULT_LAME_MP3_QUALITY)
        // Create and run thread used to encode data
        // The thread will
        mEncodeThread = DataEncodeThread(mRecordFile, mBufferSize)
        mEncodeThread?.start()
        mAudioRecord?.setRecordPositionUpdateListener(mEncodeThread,
                mEncodeThread?.handler)
        mAudioRecord?.positionNotificationPeriod = FRAME_COUNT
    }

    companion object {
        // =======================AudioRecord Default
        // Settings=======================
        private val DEFAULT_AUDIO_SOURCE = MediaRecorder.AudioSource.MIC
        /**
         * 以下三项为默认配置参数。Google Android文档明确表明只有以下3个参数是可以在所有设备上保证支持的。
         */
        val DEFAULT_SAMPLING_RATE = 44100// 模拟器仅支持从麦克风输入8kHz采样率
        private val DEFAULT_CHANNEL_CONFIG = AudioFormat.CHANNEL_IN_MONO
        /**
         * 下面是对此的封装 private static final int DEFAULT_AUDIO_FORMAT =
         * AudioFormat.ENCODING_PCM_16BIT;
         */
        private val DEFAULT_AUDIO_FORMAT = PCMFormat.PCM_16BIT

        // ======================Lame Default Settings=====================
        val DEFAULT_LAME_MP3_QUALITY = 4
        /**
         * 与DEFAULT_CHANNEL_CONFIG相关，因为是mono单声，所以是1
         */
        val DEFAULT_LAME_IN_CHANNEL = 1
        /**
         * Encoded bit rate. MP3 file will be encoded with bit rate 32kbps
         */
        val DEFAULT_LAME_MP3_BIT_RATE = 32

        // ==================================================================

        /**
         * 自定义 每160帧作为一个周期，通知一下需要进行编码
         */
        private val FRAME_COUNT = 160

        private val MAX_VOLUME = 2000
    }
}